File: ui/src/lib/mcp-server-resolver.ts
export const CANONICAL_JIRA_MCP_NAME = 'jira'
export const CANONICAL_GITLAB_MCP_NAME = 'gitlab'
export const CANONICAL_GITHUB_MCP_NAME = 'github'

export type McpServerMap = Record<string, Record<string, unknown>>

function argsBlob(config: Record<string, unknown>): string {
  const args = config.args
  if (!Array.isArray(args)) {
    return ''
  }
  return args.map(value => String(value)).join(' ').toLowerCase()
}

function envKeys(config: Record<string, unknown>): Set<string> {
  const env = config.env
  if (!env || typeof env !== 'object' || Array.isArray(env)) {
    return new Set()
  }
  return new Set(Object.keys(env as Record<string, unknown>).map(key => key.toUpperCase()))
}

export function isJiraMcpServer(name: string, config: Record<string, unknown>): boolean {
  const lowered = name.trim().toLowerCase()
  if (lowered === CANONICAL_JIRA_MCP_NAME || lowered === 'atlassian' || lowered.includes('jira') || lowered.includes('atlassian')) {
    return true
  }
  const args = argsBlob(config)
  if (args.includes('mcp-atlassian')) {
    return true
  }
  if (args.includes('mcp-remote') && args.includes('atlassian')) {
    return true
  }
  const env = envKeys(config)
  return env.has('JIRA_URL') || env.has('JIRA_USERNAME')
}

export function isGitlabMcpServer(name: string, config: Record<string, unknown>): boolean {
  const lowered = name.trim().toLowerCase()
  if (lowered === CANONICAL_GITLAB_MCP_NAME || lowered.includes('gitlab')) {
    return true
  }
  const args = argsBlob(config)
  return (
    args.includes('server-gitlab') ||
    args.includes('@modelcontextprotocol/server-gitlab') ||
    envKeys(config).has('GITLAB_API_URL') ||
    envKeys(config).has('GITLAB_PERSONAL_ACCESS_TOKEN')
  )
}

export function isGithubMcpServer(name: string, config: Record<string, unknown>): boolean {
  const lowered = name.trim().toLowerCase()
  if (lowered === CANONICAL_GITHUB_MCP_NAME || lowered.includes('github')) {
    return true
  }
  const args = argsBlob(config)
  return (
    args.includes('server-github') ||
    args.includes('@modelcontextprotocol/server-github') ||
    envKeys(config).has('GITHUB_PERSONAL_ACCESS_TOKEN')
  )
}

function resolveServerName(
  servers: McpServerMap,
  canonical: string,
  predicate: (name: string, config: Record<string, unknown>) => boolean
): string | null {
  const canonicalConfig = servers[canonical]
  if (canonicalConfig && predicate(canonical, canonicalConfig)) {
    return canonical
  }

  for (const [name, config] of Object.entries(servers)) {
    if (config && predicate(name, config)) {
      return name
    }
  }

  return null
}

export function resolveJiraMcpServer(servers: McpServerMap): { name: string; config: Record<string, unknown> } | null {
  const name = resolveServerName(servers, CANONICAL_JIRA_MCP_NAME, isJiraMcpServer)
  if (!name) {
    return null
  }
  return { name, config: servers[name] ?? {} }
}

export function resolveGitlabMcpServer(servers: McpServerMap): { name: string; config: Record<string, unknown> } | null {
  const name = resolveServerName(servers, CANONICAL_GITLAB_MCP_NAME, isGitlabMcpServer)
  if (!name) {
    return null
  }
  return { name, config: servers[name] ?? {} }
}

export function resolveGithubMcpServer(servers: McpServerMap): { name: string; config: Record<string, unknown> } | null {
  const name = resolveServerName(servers, CANONICAL_GITHUB_MCP_NAME, isGithubMcpServer)
  if (!name) {
    return null
  }
  return { name, config: servers[name] ?? {} }
}

export function readMcpServers(config: { mcp_servers?: unknown } | null | undefined): McpServerMap {
  const servers = config?.mcp_servers
  if (!servers || typeof servers !== 'object' || Array.isArray(servers)) {
    return {}
  }
  return servers as McpServerMap
}

File: lc_server/integrations/mcp_server_resolver.py
"""Resolve Hermes MCP server names from user config (not only canonical jira/gitlab/github)."""

from __future__ import annotations

from typing import Any

CANONICAL_JIRA_MCP_NAME = "jira"
CANONICAL_GITLAB_MCP_NAME = "gitlab"
CANONICAL_GITHUB_MCP_NAME = "github"


def _args_blob(config: dict[str, Any]) -> str:
    args = config.get("args") or []
    if not isinstance(args, list):
        return ""
    return " ".join(str(item) for item in args).lower()


def _env_keys(config: dict[str, Any]) -> set[str]:
    env = config.get("env")
    if not isinstance(env, dict):
        return set()
    return {str(key).upper() for key in env.keys()}


def is_jira_mcp_server(name: str, config: dict[str, Any]) -> bool:
    lowered = (name or "").strip().lower()
    if lowered in {CANONICAL_JIRA_MCP_NAME, "atlassian"} or "jira" in lowered or "atlassian" in lowered:
        return True
    args = _args_blob(config)
    if "mcp-atlassian" in args:
        return True
    if "mcp-remote" in args and "atlassian" in args:
        return True
    env = _env_keys(config)
    return "JIRA_URL" in env or "JIRA_USERNAME" in env


def is_gitlab_mcp_server(name: str, config: dict[str, Any]) -> bool:
    lowered = (name or "").strip().lower()
    if lowered == CANONICAL_GITLAB_MCP_NAME or "gitlab" in lowered:
        return True
    args = _args_blob(config)
    if "server-gitlab" in args or "@modelcontextprotocol/server-gitlab" in args:
        return True
    return "GITLAB_API_URL" in _env_keys(config) or "GITLAB_PERSONAL_ACCESS_TOKEN" in _env_keys(config)


def is_github_mcp_server(name: str, config: dict[str, Any]) -> bool:
    lowered = (name or "").strip().lower()
    if lowered == CANONICAL_GITHUB_MCP_NAME or "github" in lowered:
        return True
    args = _args_blob(config)
    if "server-github" in args or "@modelcontextprotocol/server-github" in args:
        return True
    return "GITHUB_PERSONAL_ACCESS_TOKEN" in _env_keys(config)


def _resolve_server_name(
    *,
    canonical: str,
    predicate,
    servers: dict[str, Any],
) -> str | None:
    if not isinstance(servers, dict):
        return None
    canonical_cfg = servers.get(canonical)
    if isinstance(canonical_cfg, dict) and predicate(canonical, canonical_cfg):
        return canonical
    for name, config in servers.items():
        if not isinstance(config, dict):
            continue
        if predicate(str(name), config):
            return str(name)
    return None


def resolve_jira_mcp_server_name(servers: dict[str, Any] | None = None) -> str | None:
    if servers is None:
        from lc_server.integrations.mcp_config_bridge import load_effective_mcp_servers

        servers = load_effective_mcp_servers()
    return _resolve_server_name(canonical=CANONICAL_JIRA_MCP_NAME, predicate=is_jira_mcp_server, servers=servers)


def resolve_gitlab_mcp_server_name(servers: dict[str, Any] | None = None) -> str | None:
    if servers is None:
        from lc_server.integrations.mcp_config_bridge import load_effective_mcp_servers

        servers = load_effective_mcp_servers()
    return _resolve_server_name(canonical=CANONICAL_GITLAB_MCP_NAME, predicate=is_gitlab_mcp_server, servers=servers)


def resolve_github_mcp_server_name(servers: dict[str, Any] | None = None) -> str | None:
    if servers is None:
        from lc_server.integrations.mcp_config_bridge import load_effective_mcp_servers

        servers = load_effective_mcp_servers()
    return _resolve_server_name(canonical=CANONICAL_GITHUB_MCP_NAME, predicate=is_github_mcp_server, servers=servers)


def active_jira_mcp_name(servers: dict[str, Any] | None = None) -> str:
    return resolve_jira_mcp_server_name(servers) or CANONICAL_JIRA_MCP_NAME


def active_gitlab_mcp_name(servers: dict[str, Any] | None = None) -> str:
    return resolve_gitlab_mcp_server_name(servers) or CANONICAL_GITLAB_MCP_NAME


def active_github_mcp_name(servers: dict[str, Any] | None = None) -> str:
    return resolve_github_mcp_server_name(servers) or CANONICAL_GITHUB_MCP_NAME
